File: Kryptos/Model/KeyArchive.swift
import Foundation
import CipherCore

struct KeyArchive: Codable {
    static let magic = "keys"
    static let version = 1
    static let minPasswordLength = 8

    var kryptos: String
    var v: Int
    var created: Int64
    var profiles: [ArchivedProfile]
    var pgpIdentities: [ArchivedPgpIdentity]
    var pgpRecipients: [ArchivedPgpRecipient]

    struct ArchivedProfile: Codable {
        var id: String
        var name: String
        var identity: String
        var registrationId: Int64
        var signedPreKeyId: Int64
        var signedPreKeyPub: String
        var signedPreKeySig: String
        var kyberPreKeyId: Int64
        var kyberPreKeyPub: String
        var kyberPreKeySig: String
        var prekeyCreatedAt: Int64?
        var nextSignedPreKeyId: Int64
        var nextKyberPreKeyId: Int64
        var nextOneTimePreKeyId: Int64
        var oneTimePreKeyIds: [Int64]
        var retired: [ArchivedRetired]
        var autoDelete: [String: Double]
        var contacts: [ArchivedContact]
        var preKeys: [String: String]
        var signedPreKeys: [String: String]
        var kyberPreKeys: [String: String]
        var sessions: [String: String]
        var identities: [String: String]
    }

    struct ArchivedRetired: Codable {
        var signedPreKeyId: Int64
        var kyberPreKeyId: Int64
        var retiredAt: Int64
    }

    struct ArchivedContact: Codable {
        var fingerprint: String
        var displayName: String
    }

    struct ArchivedPgpIdentity: Codable {
        var id: String
        var name: String
        var email: String
        var fingerprint: String
        var algo: String
        var created: Int64
        var publicKey: String
        var secret: String
    }

    struct ArchivedPgpRecipient: Codable {
        var name: String
        var publicKey: String
        var fingerprint: String
    }

    var isEmpty: Bool { profiles.isEmpty && pgpIdentities.isEmpty && pgpRecipients.isEmpty }

    var contactCount: Int { profiles.reduce(0) { $0 + $1.contacts.count } }
}

enum KeyArchiveError: LocalizedError {
    case passwordTooShort
    case unreadable
    case nothingToExport
    case writeFailed

    var errorDescription: String? {
        switch self {
        case .passwordTooShort:
            return String(localized: "Use at least 8 characters for the backup password.")
        case .unreadable:
            return String(localized: "Could not read this backup — wrong password, or the file is not a Kryptos key backup.")
        case .nothingToExport:
            return String(localized: "There are no keys to export yet.")
        case .writeFailed:
            return String(localized: "Could not create the backup file.")
        }
    }
}

extension KeyArchive {
    static func make(profiles: [ArchivedProfile],
                     pgpIdentities: [ArchivedPgpIdentity],
                     pgpRecipients: [ArchivedPgpRecipient]) -> KeyArchive {
        KeyArchive(kryptos: magic, v: version,
                   created: Int64(Date().timeIntervalSince1970 * 1000),
                   profiles: profiles, pgpIdentities: pgpIdentities, pgpRecipients: pgpRecipients)
    }

    func sealed(password: String) throws -> String {
        guard password.count >= KeyArchive.minPasswordLength else { throw KeyArchiveError.passwordTooShort }
        guard !isEmpty else { throw KeyArchiveError.nothingToExport }
        var plain = [UInt8](try JSONEncoder().encode(self))
        defer { Argon2id.zero(&plain) }
        let raw = try PasswordCipher.encrypt(Data(plain), password: password)
        return WireFormat.token(raw)
    }

    static func opened(_ text: String, password: String) throws -> KeyArchive {
        guard let raw = WireFormat.tokenBytes(text.trimmingCharacters(in: .whitespacesAndNewlines)),
              let plain = try? PasswordCipher.decrypt(raw, password: password),
              let archive = try? JSONDecoder().decode(KeyArchive.self, from: plain),
              archive.kryptos == magic, archive.v == version else { throw KeyArchiveError.unreadable }
        guard !archive.isEmpty else { throw KeyArchiveError.unreadable }
        return archive
    }
}

File: android/app/src/main/java/com/kryptos/android/core/KeyArchive.kt
package com.kryptos.android.core

import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.decodeFromStream
import kotlinx.serialization.json.encodeToStream

@Serializable
data class ArchivedContact(val fingerprint: String, val displayName: String)

@Serializable
data class ArchivedRetired(val signedPreKeyId: Long, val kyberPreKeyId: Long, val retiredAt: Long)

@Serializable
data class ArchivedProfile(
    val id: String,
    val name: String,
    val identity: String,
    val registrationId: Long,
    val signedPreKeyId: Long,
    val signedPreKeyPub: String,
    val signedPreKeySig: String,
    val kyberPreKeyId: Long,
    val kyberPreKeyPub: String,
    val kyberPreKeySig: String,
    val prekeyCreatedAt: Long? = null,
    val nextSignedPreKeyId: Long,
    val nextKyberPreKeyId: Long,
    val nextOneTimePreKeyId: Long,
    val oneTimePreKeyIds: List<Long> = emptyList(),
    val retired: List<ArchivedRetired> = emptyList(),
    val autoDelete: Map<String, Double> = emptyMap(),
    val contacts: List<ArchivedContact> = emptyList(),
    val preKeys: Map<String, String> = emptyMap(),
    val signedPreKeys: Map<String, String> = emptyMap(),
    val kyberPreKeys: Map<String, String> = emptyMap(),
    val sessions: Map<String, String> = emptyMap(),
    val identities: Map<String, String> = emptyMap(),
)

@Serializable
data class ArchivedPgpIdentity(
    val id: String,
    val name: String,
    val email: String,
    val fingerprint: String,
    val algo: String,
    val created: Long,
    val publicKey: String,
    val secret: String,
)

@Serializable
data class ArchivedPgpRecipient(val name: String, val publicKey: String, val fingerprint: String)

@Serializable
data class KeyArchive(
    val kryptos: String = MAGIC,
    val v: Int = VERSION,
    val created: Long = System.currentTimeMillis(),
    val profiles: List<ArchivedProfile> = emptyList(),
    val pgpIdentities: List<ArchivedPgpIdentity> = emptyList(),
    val pgpRecipients: List<ArchivedPgpRecipient> = emptyList(),
) {
    val isEmpty: Boolean get() = profiles.isEmpty() && pgpIdentities.isEmpty() && pgpRecipients.isEmpty()

    val contactCount: Int get() = profiles.sumOf { it.contacts.size }

    companion object {
        const val MAGIC = "keys"
        const val VERSION = 1
        const val MIN_PASSWORD_LENGTH = 8
        const val MAX_FILE_BYTES = 8 * 1024 * 1024

        private val json = Json { ignoreUnknownKeys = true; encodeDefaults = true }

        private class WipingBuffer : java.io.OutputStream() {
            private var buf = ByteArray(8192)
            private var size = 0

            private fun ensure(extra: Int) {
                if (size + extra <= buf.size) return
                var capacity = buf.size
                while (capacity < size + extra) capacity *= 2
                val grown = buf.copyOf(capacity)
                buf.fill(0)
                buf = grown
            }

            override fun write(b: Int) {
                ensure(1)
                buf[size++] = b.toByte()
            }

            override fun write(b: ByteArray, off: Int, len: Int) {
                ensure(len)
                b.copyInto(buf, size, off, off + len)
                size += len
            }

            fun drain(): ByteArray {
                val out = buf.copyOf(size)
                buf.fill(0)
                size = 0
                return out
            }
        }

        @OptIn(ExperimentalSerializationApi::class)
        fun seal(archive: KeyArchive, password: String): String {
            if (password.length < MIN_PASSWORD_LENGTH) throw ArchiveException(ArchiveException.Kind.PASSWORD_TOO_SHORT)
            if (archive.isEmpty) throw ArchiveException(ArchiveException.Kind.NOTHING_TO_EXPORT)
            val buffer = WipingBuffer()
            val plain = try {
                json.encodeToStream(serializer(), archive, buffer)
                buffer.drain()
            } catch (e: Exception) {
                buffer.drain().fill(0)
                throw e
            }
            try {
                return WireFormat.token(PasswordCipher.encrypt(plain, password))
            } finally {
                plain.fill(0)
            }
        }

        @OptIn(ExperimentalSerializationApi::class)
        fun open(text: String, password: String): KeyArchive {
            val raw = WireFormat.tokenBytes(text.trim())
                ?: throw ArchiveException(ArchiveException.Kind.UNREADABLE)
            val plain = try {
                PasswordCipher.decrypt(raw, password)
            } catch (e: Exception) {
                throw ArchiveException(ArchiveException.Kind.UNREADABLE)
            }
            val archive = try {
                json.decodeFromStream<KeyArchive>(java.io.ByteArrayInputStream(plain))
            } catch (e: Exception) {
                throw ArchiveException(ArchiveException.Kind.UNREADABLE)
            } finally {
                plain.fill(0)
            }
            if (archive.kryptos != MAGIC || archive.v != VERSION || archive.isEmpty) {
                throw ArchiveException(ArchiveException.Kind.UNREADABLE)
            }
            return archive
        }
    }
}

class ArchiveException(val kind: Kind) : Exception(kind.name) {
    enum class Kind { PASSWORD_TOO_SHORT, UNREADABLE, NOTHING_TO_EXPORT, WRITE_FAILED }
}
